File: src/state/selectors.ts
// =============================================================================
// selectors.ts — Pure derived-state selectors for AppState
// =============================================================================
//
// Selectors are pure functions that compute derived values from the AppState
// atom. They encapsulate common lookup + validation patterns so that multiple
// consumers (React components, REPL logic, input routing) share one canonical
// implementation rather than duplicating task-lookup boilerplate.
//
// Design principles:
//   1. Pure — no side effects, no mutations, no I/O.
//   2. Minimal — accept only the AppState slices they need (via Pick<>),
//      so callers can pass partial mocks in tests.
//   3. Type-safe — return discriminated unions so callers can narrow with
//      pattern matching and get exhaustive checking.
//
// These selectors are NOT React hooks. They can be used in any context:
//   • Inside useAppState(selector) for reactive rendering
//   • In imperative code with store.getState()
//   • In tests with plain objects
// =============================================================================

/**
 * Selectors for deriving computed state from AppState.
 * Keep selectors pure and simple - just data extraction, no side effects.
 */

import type { InProcessTeammateTaskState } from '../tasks/InProcessTeammateTask/types.js'
import { isInProcessTeammateTask } from '../tasks/InProcessTeammateTask/types.js'
import type { LocalAgentTaskState } from '../tasks/LocalAgentTask/LocalAgentTask.js'
import type { AppState } from './AppStateStore.js'

/**
 * Get the currently viewed teammate task, if any.
 * Returns undefined if:
 * - No teammate is being viewed (viewingAgentTaskId is undefined)
 * - The task ID doesn't exist in tasks
 * - The task is not an in-process teammate task
 */
// ---------------------------------------------------------------------------
// getViewedTeammateTask — Resolves the task object for the teammate currently
// being viewed in the UI.
//
// This is a 3-step validation pipeline:
//   1. Is viewingAgentTaskId set? (are we viewing anyone?)
//   2. Does the ID map to an existing task? (guard against stale IDs)
//   3. Is that task an in-process teammate? (not a local-agent or other type)
//
// Returns the typed InProcessTeammateTaskState or undefined.
// Uses Pick<AppState, …> so callers only need to provide the two fields,
// making this easy to test with minimal mocks.
// ---------------------------------------------------------------------------
export function getViewedTeammateTask(
  appState: Pick<AppState, 'viewingAgentTaskId' | 'tasks'>,
): InProcessTeammateTaskState | undefined {
  const { viewingAgentTaskId, tasks } = appState

  // Not viewing any teammate
  if (!viewingAgentTaskId) {
    return undefined
  }

  // Look up the task
  const task = tasks[viewingAgentTaskId]
  if (!task) {
    return undefined
  }

  // Verify it's an in-process teammate task
  if (!isInProcessTeammateTask(task)) {
    return undefined
  }

  return task
}

/**
 * Return type for getActiveAgentForInput selector.
 * Discriminated union for type-safe input routing.
 */
// ---------------------------------------------------------------------------
// ActiveAgentForInput — Discriminated union describing where user input
// should be directed. The `type` discriminant enables exhaustive switch/case:
//   - 'leader':      Input goes to the main leader agent (default)
//   - 'viewed':      Input goes to the in-process teammate being viewed
//   - 'named_agent': Input goes to a local agent task (non-teammate subagent)
// ---------------------------------------------------------------------------
export type ActiveAgentForInput =
  | { type: 'leader' }
  | { type: 'viewed'; task: InProcessTeammateTaskState }
  | { type: 'named_agent'; task: LocalAgentTaskState }

/**
 * Determine where user input should be routed.
 * Returns:
 * - { type: 'leader' } when not viewing a teammate (input goes to leader)
 * - { type: 'viewed', task } when viewing an agent (input goes to that agent)
 *
 * Used by input routing logic to direct user messages to the correct agent.
 */
// ---------------------------------------------------------------------------
// getActiveAgentForInput — Determines the target agent for the next user input.
//
// Resolution order:
//   1. If viewing an in-process teammate → route to that teammate
//   2. If viewing a local agent task → route to that named agent
//   3. Otherwise → route to the leader (default behavior)
//
// This selector drives the input routing in the REPL: when a user types a
// message, the system needs to know whether to send it to the leader, a
// viewed teammate, or a named subagent. The resolution priority ensures
// that explicitly viewed agents take precedence.
// ---------------------------------------------------------------------------
export function getActiveAgentForInput(
  appState: AppState,
): ActiveAgentForInput {
  // First: check if we're viewing an in-process teammate
  const viewedTask = getViewedTeammateTask(appState)
  if (viewedTask) {
    return { type: 'viewed', task: viewedTask }
  }

  // Second: check if we're viewing a local agent task (non-teammate subagent)
  const { viewingAgentTaskId, tasks } = appState
  if (viewingAgentTaskId) {
    const task = tasks[viewingAgentTaskId]
    if (task?.type === 'local_agent') {
      return { type: 'named_agent', task }
    }
  }

  // Default: input goes to the leader
  return { type: 'leader' }
}

File: python_rewrite/claude_code/state/selectors.py
"""
Pure derived-state selectors for AppState.

Selectors are pure functions that compute derived values from the AppState
atom.  They encapsulate common lookup + validation patterns so that multiple
consumers (TUI components, REPL logic, input routing) share one canonical
implementation rather than duplicating task-lookup boilerplate.

Design principles:
    1. Pure -- no side effects, no mutations, no I/O.
    2. Minimal -- accept only the AppState slices they need (via Protocol /
       TypedDict), so callers can pass partial mocks in tests.
    3. Type-safe -- return discriminated unions so callers can narrow with
       pattern matching and get exhaustive checking.

These selectors are NOT framework hooks.  They can be used in any context:
    - Inside reactive subscriptions for rendering
    - In imperative code with ``store.get_state()``
    - In tests with plain objects
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal, Protocol, runtime_checkable

from .app_state import AppState, PermissionMode, ToolPermissionContext


# ---------------------------------------------------------------------------
# Task type guards
# ---------------------------------------------------------------------------

def is_in_process_teammate_task(task: dict[str, Any]) -> bool:
    """Check whether a task dict represents an in-process teammate task."""
    return task.get("type") == "in_process_teammate"


def is_local_agent_task(task: dict[str, Any]) -> bool:
    """Check whether a task dict represents a local agent task."""
    return task.get("type") == "local_agent"


# ---------------------------------------------------------------------------
# ActiveAgentForInput -- Discriminated union describing where user input
# should be directed.
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class ActiveAgentLeader:
    """Input goes to the main leader agent (default)."""
    type: Literal["leader"] = "leader"


@dataclass(frozen=True)
class ActiveAgentViewed:
    """Input goes to the in-process teammate being viewed."""
    type: Literal["viewed"] = "viewed"
    task: dict[str, Any] = None  # type: ignore[assignment]


@dataclass(frozen=True)
class ActiveAgentNamed:
    """Input goes to a local agent task (non-teammate subagent)."""
    type: Literal["named_agent"] = "named_agent"
    task: dict[str, Any] = None  # type: ignore[assignment]


ActiveAgentForInput = ActiveAgentLeader | ActiveAgentViewed | ActiveAgentNamed


# ---------------------------------------------------------------------------
# get_viewed_teammate_task
# ---------------------------------------------------------------------------

def get_viewed_teammate_task(
    app_state: AppState,
) -> dict[str, Any] | None:
    """Get the currently viewed teammate task, if any.

    Returns ``None`` if:
        - No teammate is being viewed (``viewing_agent_task_id`` is None)
        - The task ID doesn't exist in tasks
        - The task is not an in-process teammate task

    This is a 3-step validation pipeline:
        1. Is viewing_agent_task_id set? (are we viewing anyone?)
        2. Does the ID map to an existing task? (guard against stale IDs)
        3. Is that task an in-process teammate? (not a local-agent or other type)
    """
    viewing_id = app_state.viewing_agent_task_id

    # Not viewing any teammate
    if viewing_id is None:
        return None

    # Look up the task
    task = app_state.tasks.get(viewing_id)
    if task is None:
        return None

    # Verify it's an in-process teammate task
    if not is_in_process_teammate_task(task):
        return None

    return task


# ---------------------------------------------------------------------------
# get_active_agent_for_input
# ---------------------------------------------------------------------------

def get_active_agent_for_input(
    app_state: AppState,
) -> ActiveAgentForInput:
    """Determine where user input should be routed.

    Resolution order:
        1. If viewing an in-process teammate -> route to that teammate
        2. If viewing a local agent task -> route to that named agent
        3. Otherwise -> route to the leader (default behaviour)

    This selector drives the input routing in the REPL: when a user types a
    message, the system needs to know whether to send it to the leader, a
    viewed teammate, or a named subagent.

    Returns:
        - ``ActiveAgentLeader()`` when not viewing a teammate
        - ``ActiveAgentViewed(task=...)`` when viewing an in-process teammate
        - ``ActiveAgentNamed(task=...)`` when viewing a local agent task
    """
    # First: check if we're viewing an in-process teammate
    viewed_task = get_viewed_teammate_task(app_state)
    if viewed_task is not None:
        return ActiveAgentViewed(task=viewed_task)

    # Second: check if we're viewing a local agent task (non-teammate subagent)
    viewing_id = app_state.viewing_agent_task_id
    if viewing_id is not None:
        task = app_state.tasks.get(viewing_id)
        if task is not None and is_local_agent_task(task):
            return ActiveAgentNamed(task=task)

    # Default: input goes to the leader
    return ActiveAgentLeader()


# ---------------------------------------------------------------------------
# get_effective_permission_mode
# ---------------------------------------------------------------------------

def get_effective_permission_mode(
    app_state: AppState,
) -> PermissionMode:
    """Get the effective permission mode, accounting for overrides.

    The effective mode resolves the current ``ToolPermissionContext.mode``
    after considering:
        - The base mode set by the user (default, plan, auto, bypass)
        - Session-level overrides (e.g. bypass permissions mode flag)

    This is the canonical way to determine the current permission mode
    for tool execution decisions.
    """
    return app_state.tool_permission_context.mode


# ---------------------------------------------------------------------------
# Convenience selectors
# ---------------------------------------------------------------------------

def get_current_model(app_state: AppState) -> str | None:
    """Get the currently configured model (alias or full name), or None."""
    return app_state.main_loop_model


def is_plan_mode(app_state: AppState) -> bool:
    """Check if the session is currently in plan mode."""
    return app_state.tool_permission_context.mode == PermissionMode.PLAN


def is_auto_mode(app_state: AppState) -> bool:
    """Check if the session is currently in auto/ungated-auto mode."""
    mode = app_state.tool_permission_context.mode
    return mode in (PermissionMode.AUTO, PermissionMode.UNGATED_AUTO)


def get_active_task_count(app_state: AppState) -> int:
    """Count tasks that are currently running (not completed/failed)."""
    return sum(
        1
        for task in app_state.tasks.values()
        if isinstance(task, dict)
        and task.get("status") in ("running", "pending")
    )


def has_active_speculation(app_state: AppState) -> bool:
    """Check if there is an active speculative execution in flight."""
    return getattr(app_state.speculation, "status", "idle") == "active"


__all__ = [
    "ActiveAgentForInput",
    "ActiveAgentLeader",
    "ActiveAgentNamed",
    "ActiveAgentViewed",
    "get_active_agent_for_input",
    "get_active_task_count",
    "get_current_model",
    "get_effective_permission_mode",
    "get_viewed_teammate_task",
    "has_active_speculation",
    "is_auto_mode",
    "is_in_process_teammate_task",
    "is_local_agent_task",
    "is_plan_mode",
]
